src/backend/lib/utils.js

```diff
@@ -0,0 +1,299 @@
+    })
+  );
+  return upgradeCandidates.sort(installerCompare).reverse()[0];
+}
+
+/**
+ * For a given filepath to an APK, get the UpgradeOption details
+ *
+ * @param {string} filepath
+ * @returns {Promise<InstallerInt>}
+ */
+async function getInstallerInfo(filepath) {
+  if (path.extname(filepath) !== ".apk")
+    throw new Error(`${filepath} is not an .apk file`);
+  const [
+    hash,
+    stat,
+    { versionName, versionCode, applicationId, arch, minSdkVersion },
+  ] = await Promise.all([
+    hasha.fromFile(filepath, { algorithm: "sha256" }),
+    fs.promises.stat(filepath),
+    getApkMetadata(filepath),
+  ]);
+  return {
+    hash,
+    hashType: "sha256",
+    versionName,
+    versionCode,
+    applicationId,
+    platform: "android",
+    arch,
+    size: stat.size,
+    minSdkVersion,
+    filepath,
+  };
+}
+
+/**
+ * Get version name, version code & applicationId from an apk file
+ *
+ * @param {string} filepath path to apk
+ * @returns {Promise<Pick<InstallerInt, 'versionName' | 'versionCode' | 'applicationId' | 'arch' | 'minSdkVersion'>>}
+ */
+async function getApkMetadata(filepath) {
+  const parser = new ApkParser(filepath);
+  const [result, arch] = await Promise.all([
+    parser.parse(),
+    getApkAbis(filepath),
+  ]);
+  return {
+    versionName: result.versionName,
+    versionCode: result.versionCode,
+    applicationId: result.package,
+    arch,
+    minSdkVersion: (result.usesSdk || {}).minSdkVersion || 1,
+  };
+}
+
+// If an APK does not have native code (e.g. no `lib` folder) then we assume it runs on these ABIs
+/** @type {InstallerInt['arch']} */
+const DEFAULT_ABIS = ["armeabi-v7a", "arm64-v8a", "x86", "x86_64"];
+
+/**
+ * Get the ABIs supported by an APK
+ *
+ * @param {string} filepath
+ * @returns {Promise<InstallerInt['arch']>}
+ */
+async function getApkAbis(filepath) {
+  const zip = new ZipFs(filepath);
+  /** @type {InstallerInt['arch']} */
+  const libDirs = (await promisify(zip.readdir.bind(zip))("lib")).filter((
+    /** @type {string} */ name
+  ) =>
+    DEFAULT_ABIS.includes(
+      // @ts-ignore - this one is a pain to type correctly!
+      name
+    )
+  );
+  await promisify(zip.close.bind(zip))();
+  return libDirs.length ? libDirs : DEFAULT_ABIS.slice(0);
+}
+
+/**
+ * Wrap an asynchronously created writable stream with another writable stream,
+ * which will run a flush function _after_ the wrapped stream has already
+ * finished, but before the returned stream will finish.
+ *
+ * Useful if you want to synchronously return a stream, then asynchronously
+ * create a write stream, then do some work after the internal stream has
+ * finished writing (e.g. move a file) before the returned stream finishes.
+ *
+ * @param {object} options
+ * @param {() => Promise<stream.Writable>} options.createStream
+ * @param {(wrappedStream: stream.Writable) => Promise<void>} options.finalize
+ * @returns {stream.Writable}
+ */
+function beforeAfterStream({ createStream, finalize }) {
```

Comment:
This is some pretty tricky code to get 100% correct! I don't even feel totally qualified to evaluate its correctness. As far as I know, the write stream handling code I had in UpgradeStorage worked correctly. Why was this new [complex!] piece of code needed?

Reply:
Yes this turned out to be more complex than I expected. I embarked upon it from a desire to create a simpler interface for UpgradeStorage, and didn't like the confusion of having to handle two callbacks - once when the write stream to tmp file finished, and once for when the file finally moves. In the end I think an interface like: `upgradeStorage.writeStream(readableStream, callback)` would have been simpler, even though it also creates a lack of clarity about where errors on `readableStream` should be handled. I lean heavily on tests here, and I copied the tests from `duplexify`. I feel confident that how we intend to use this helper, and the errors that we might get, are covered in the tests.